### PR TITLE
upgrade socket engine

### PR DESCRIPTION
**Pull request by urbanishimwe**
the new socket implementations uses `T_PACKETV3`
go version: `go1.16.2 linux/amd64`
cpu: `Intel(R) Core(TM) i5-4300U CPU @ 1.90GHz`

socket benchmark: `sudo go test -cpu=1 -benchtime=100x -bench ReadSocket -run ^$ github.com/buger/goreplay/capture`
```
BenchmarkReadSocketBufferFull 	     100	    970767 ns/op	      1223 packets/op
```
pcap benchmark: `sudo go test -cpu=1 -benchtime=100x -bench ReadPcap -run ^$ github.com/buger/goreplay/capture`
```
BenchmarkReadPcapBufferFull 	     100	   2211748 ns/op	      1000 packets/op
```

### CODE REVIEW

File: capture/capture.go [hunk 1]
```diff
@@ -271,27 +253,39 @@ func (l *Listener) PcapHandle(ifi net.Interface) (handle *pcap.Handle, err error
...
-		handle.Close()
+	snaplen := 128 * 1024
+	filter, err := pcap.CompileBPFFilter(layers.LinkTypeEthernet, snaplen, l.BPFFilter)
+	if err != nil {
 		return nil, fmt.Errorf("BPF filter error: %q%s, interface: %q", err, l.BPFFilter, ifi.Name)
 	}
-	handle.SetLoopbackIndex(int32(l.loopIndex))
-	return
+	f := make([]bpf.RawInstruction, len(filter))
+	for i := range filter {
```
Comment (buger) [inline capture/capture.go:281]: What this lines do?

File: capture/capture.go [hunk 2]
```diff
@@ -271,27 +253,39 @@ func (l *Listener) PcapHandle(ifi net.Interface) (handle *pcap.Handle, err error
...
 	if l.BPFFilter != "" {
 		if l.BPFFilter[0] != '(' || l.BPFFilter[len(l.BPFFilter)-1] != ')' {
 			l.BPFFilter = "(" + l.BPFFilter + ")"
 		}
 	} else {
 		l.BPFFilter = l.Filter(ifi)
 	}
-	if err = handle.SetBPFFilter(l.BPFFilter); err != nil {
-		handle.Close()
+	snaplen := 128 * 1024
```
Comment (buger) [inline capture/capture.go:275]: Why you picked this value?

File: capture/capture.go [hunk 3]
```diff
@@ -311,6 +304,12 @@ func (l *Listener) read(handler Handler) {
 					}
 					return // can't find the linktype size
 				}
+			} else {
+				linkType = int(hndl.(*socket).h.LinkType())
+				linkSize = 14
+				if hndl.(*socket).isCooked() {
```
Comment (buger) [inline capture/capture.go:310]: What is cooked mean? 
Can you add a comment?

File: capture/capture.go (hunk 2 above)
Comment (urbanishimwe) [reply capture/capture.go:275]: It's unlikely that packet will have length greater than that value!

File: capture/capture.go (hunk 1 above)
Comment (urbanishimwe) [reply capture/capture.go:281]: convert `[]pcap.BPFInstruction` to `[]bpf.RawInstruction`

File: capture/capture.go (hunk 3 above)
Comment (urbanishimwe) [reply capture/capture.go:310]: it checks if link layer header is removed off of the packet buffer.(`by default we do this`)

File: capture/capture.go (hunk 2 above)
Comment (buger) [reply capture/capture.go:275]: From what I remember spaplen is configured with configuration variable. Maybe also use it here?

**Pull request closed without merging**